fix(core): add auth provider to AuthRule

rjuliano opened the pull request:
*Issue #, if available:*

*Description of changes:*
This PR adds a field to represent the auth provider for an `AuthRule`. 

Currently, the Android generated model does not have this information, which will be needed for multi-auth support. This is intended to be a non-breaking change. Since the CLI does not generate this field yet, the default behavior is to use the default provider for the given auth strategy. 

I considered creating an enum to represent this, but decided against it for several reasons. The enum we currently use to represent an auth provider is actually called `AuthorizationType` and it lives in the `aws-api-appsync` module. Moving that to core doesn't make much sense because it's specific to AppSync. Creating a new enum would probably cause more confusion than what it's worth. We already have an enum called `AuthProvider` that is used to represent social sign-in providers.

Given the fact that the values that will be used to set this field are derived from the `@auth` directive (which already validates for valid entries), creating an enum just didn't seem worth it at this point.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

raphkim commented on core/src/main/java/com/amplifyframework/core/model/AuthStrategy.java:
```diff
@@ -25,23 +25,33 @@
+        this.defaultAuthProvider = defaultAuthProvider;
+    }
+
+    String getDefaultAuthProvider() {
+        return defaultAuthProvider;
+    }
```
A switch statement would be appropriate here no? Using the default parameter for this doesn't sit right with me.

richardmcclellan replied:
Can you elaborate @raphkim?  What's your concern with the implementation?

rjuliano replied:
Not sure if this makes a difference here, but default in this context means the default auth provider for a given strategy. 

Also, I'm going to try to implement this as an enum, so it might look a little bit different.

rjuliano replied:
If you look in the `AuthorizationType.from(AuthRule authRuleAnnotation)` method, you'll see what my thinking was behind adding the default provider as a property of the AuthStrategy enum.

raphkim reviewed (approved):
Looks good besides a minor nitpick

richardmcclellan reviewed (commented):
Hmm, like you, I'm a bit conflicted about whether to use an enum or String to represent the auth strategy.  I think I'm leaning more towards representing it as an enum, because that's how it is defined on the CLI side (See [@auth rule definition](https://docs.amplify.aws/cli/graphql-transformer/auth#definition)).  

I do share your concern though that it _sort of_ has some AWS specific concepts (`AWS_IAM`, `AMAZON_COGNITO_USER_POOLS`, etc), which would be leaking into core.  I think they could be generalized though, by renaming the enum case values (e.g. just `IAM`, and `USER_POOLS`).  Basically, we could match the naming that exists in the CLI definition of the auth directive.   After making it more general, I think it would be okay to move `AuthorizationType` into `core`.  What do you think?

rjuliano commented on core/src/main/java/com/amplifyframework/core/model/ModelSchema.java:
```diff
@@ -161,6 +161,21 @@ public static ModelSchema fromModelClass(@NonNull Class<? extends Model> clazz)
+     */
+    public List<AuthRule> getApplicableRules(ModelOperation operationType) {
+        List<AuthRule> result = new ArrayList<>();
+        for (AuthRule rule : authRules) {
```
That's a good point. I'm going to add that.

rjuliano replied:
This has been addressed.

rjuliano commented:
> Hmm, like you, I'm a bit conflicted about whether to use an enum or String to represent the auth strategy. I think I'm leaning more towards representing it as an enum, because that's how it is defined on the CLI side (See [@auth rule definition](https://docs.amplify.aws/cli/graphql-transformer/auth#definition)).

I see your point. Let me put something together using an enum to see what it looks like.

> 
> I do share your concern though that it _sort of_ has some AWS specific concepts (`AWS_IAM`, `AMAZON_COGNITO_USER_POOLS`, etc), which would be leaking into core. I think they could be generalized though, by renaming the enum case values (e.g. just `IAM`, and `USER_POOLS`). Basically, we could match the naming that exists in the CLI definition of the auth directive. After making it more general, I think it would be okay to move `AuthorizationType` into `core`. What do you think?

Yeah, the naming is what's tripping me up. Like I said, let me see what I can come up with.

rjuliano commented:
Ok. Here's what I did:

- Moved `AuthorizationType` to core
- Added a `DEFAULT` value to the `AuthorizationType` enum
- Modified the `AuthRule` annotation to default the `authProvider` field to `AuthorizationType.DEFAULT`
- Modified teh `AuthRule` class to use the enum instead.

rjuliano commented on core/src/main/java/com/amplifyframework/core/model/annotations/AuthRule.java:
```diff
@@ -46,6 +47,14 @@
+     * for the given AuthStrategy should be used.
+     * @return The name of the auth provider.
+     */
+    AuthorizationType provider() default AuthorizationType.DEFAULT;
```
Yeah, I agree. I think the annotation interface should be a string, but we can model the AuthRule class to use the enum. That will get rid of this ugliness.

rjuliano replied:
Updated the AuthRule annotation class to use a string for provider and created a function in AuthorizationType to translate the string values in the an enum item (or throw an exception)

rjuliano commented on core/src/main/java/com/amplifyframework/core/model/AuthorizationType.java:
```diff
@@ -55,7 +55,12 @@
      * No authorization.
      */
-    NONE;
+    NONE,
```
We do reference it in code in a few place:

```
Field
    NONE
Found usages  (3 usages found)
    Production  (3 usages found)
        Value read  (3 usages found)
            amplify-android-prs.aws-api  (3 usages found)
                com.amplifyframework.api.aws  (2 usages found)
                    AWSApiPlugin  (1 usage found)
                        configure(JSONObject, Context)  (1 usage found)
                            152 if (apiConfiguration.getAuthorizationType() != AuthorizationType.NONE) {
                    SubscriptionAuthorizer  (1 usage found)
                        createHeaders(GraphQLRequest<?>, boolean)  (1 usage found)
                            109 case NONE:
                com.amplifyframework.api.aws.auth  (1 usage found)
                    AuthRuleRequestDecorator  (1 usage found)
                        getAuthToken(AuthorizationType)  (1 usage found)
                            240 case NONE:
```

That being said, I'm not sure it's possible to create an AppSync API without at least the default auth type.

rjuliano commented on core/src/main/java/com/amplifyframework/core/model/ModelSchema.java:
```diff
@@ -161,6 +162,27 @@ public static ModelSchema fromModelClass(@NonNull Class<? extends Model> clazz)
+     * @param modelOperation The desired operation type (read, create, update, delete).
+     * @return A list of {@link AuthRule}s for the given operation.
+     */
+    public List<AuthRule> getApplicableRules(ModelOperation modelOperation) {
```
I initially was returning just a list of auth types, but there are instances where we need to know whether userPools is being used with an owner rule or private rule for example.

For example, when establishing subscriptions, if the only READ permission is tied to an owner-based rule, we have to decorate the subscription request body with the owner. But, if there is an auth rule that allows READs with the private strategy, then I will still need to use userPools, but won't decorate the request body with the ownerId.

Regarding the field-level use case, that's definitely a gap I have to close to take into account the default auth type if no auth rule is specified at the object level.

rjuliano replied:
Quick follow up on the 

> Hmm, like you, I'm a bit conflicted about whether to use an enum or String to represent the auth strategy. I think I'm leaning more towards representing it as an enum, because that's how it is defined on the CLI side (See [@auth rule definition](https://docs.amplify.aws/cli/graphql-transformer/auth#definition)).
> 
> I do share your concern though that it _sort of_ has some AWS specific concepts (`AWS_IAM`, `AMAZON_COGNITO_USER_POOLS`, etc), which would be leaking into core. I think they could be generalized though, by renaming the enum case values (e.g. just `IAM`, and `USER_POOLS`). Basically, we could match the naming that exists in the CLI definition of the auth directive. After making it more general, I think it would be okay to move `AuthorizationType` into `core`. What do you think?

Quick follow up on this. The values that the CLI puts in the config file reflect the values used by the [AppSync cloudformation resource](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-appsync-graphqlapi.html#cfn-appsync-graphqlapi-authenticationtype), which match the values we have now. With that said, I'm going to leave those names as is for now.
